api/client/samples/similar_regions_Frechet/sr.py

```diff
@@ -0,0 +1,560 @@
+
+        self.client = BatchClient(API_HOST, ACCESS_TOKEN)
+        
+        ################ Dealing with region information (load from cache or API) ################
+        self.region_info = {}
+        info_path = os.path.join(self.data_dir, REGION_INFO_CACHE)
+        if os.path.isfile(info_path):
+            self._logger.info("Reading region info from cache {}".format(info_path))
+            with open(info_path, 'rb') as f:
+                self.region_info = pickle.load(f)
+        if region_info_reload or search_region not in self.region_info:
+            ri = [self.client.lookup('regions', search_region)] # include top-level region itself
+            for l in range(5,ri[0]['level'],-1):
+                self._logger.info("Loading region info for geo level {}".format(l))
+                ri += self.client.get_descendant_regions(search_region,descendant_level=l,
+                                                              include_historical=False,
+                                                              include_details=True)
+            #reformat as dict with region_id as key
+            self.region_info = dict(zip([item['id'] for item in ri], ri))
+
+            # immediately save region_info
+            # might overwrite a large cache if it was storing "parallel" region (cache was Asia but asking for US)
+            with open(info_path, 'wb') as f:
+                self._logger.info("Saving region info file {}".format(info_path))
+                pickle.dump(self.region_info,f)    
+        else:
+            # We might have too many regions in the info cache - this happens if cache was filled for a parent
+            # of region currently requested (filled for entire world but we want only US, for example)
+            top_info = self.region_info[search_region]
+            self._logger.info("Search region is present in cache of size {}".format(len(self.region_info)))
+            level = top_info['level'] # can only have 2/3/4/5
+            to_include = [search_region]
+            ri = {search_region:top_info}
+            while level<5:
+                to_include = [r for r in [j for r in to_include for j in self.region_info[r]['contains']] 
+                              if self.region_info.get(r,{'level':-1})['level']==level+1]
+                ri.update({r:self.region_info[r] for r in to_include})
+                level += 1
+            self.region_info = ri
+            self._logger.info("Trimmed region info to {} items".format(len(self.region_info)))
+        self.needed_regions = sorted(self.region_info.keys())
+        self.needed_properties = sorted(self.metric_instance.keys())
+        
+        ##################### Data download ################################################
```

Comment:
let's make each of these clearly separated sections a function call... 

-  self. _initialize_regions()
-  self._data_download()
-  self._normalize()
etc.

Reply:
For "single-use" code I personally do prefer a single stream of code with comments rather than a sequence of calls to single-use functions but yes, this is somewhat contrary to current software engineering "wisdom", so as you wish ...

Reply:
A good rule of thumb is that you should be able to read the whole function without too much scrolling, so you can understand what it does and not miss any side effects etc. It's also good to be able to unit test each logical block.  Single use becomes n-uses when you want to test :)